repository/src/main/java/org/apache/atlas/accesscontrol/purpose/AtlasPurposeService.java

```diff
@@ -0,0 +1,546 @@
+
+        if(!purpose.getTypeName().equals(PURPOSE_ENTITY_TYPE)) {
+            throw new AtlasBaseException(BAD_REQUEST, "Please provide entity of type " + PURPOSE_ENTITY_TYPE);
+        }
+
+        if(!purpose.getStatus().equals(AtlasEntity.Status.ACTIVE)) {
+            LOG.info("Purpose with guid {} is already deleted/purged", purpose.getGuid());
+            return;
+        }
+
+        cleanRangerPolicies(purpose);
+
+        aliasStore.deleteAlias(getESAliasName(purpose));
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private void cleanRangerPolicies(AtlasEntity purpose) throws AtlasBaseException {
+        List<RangerPolicy> rangerPolicies = fetchRangerPoliciesByLabel(atlasRangerService,
+                "tag",
+                null,
+                getPurposeLabel(purpose.getGuid()));
+
+        if (CollectionUtils.isEmpty(rangerPolicies)) {
+            rangerPolicies = fetchRangerPoliciesByResourcesForPurposeDelete(atlasRangerService, purpose);
+        }
+
+        if (CollectionUtils.isNotEmpty(rangerPolicies)) {
+            for (RangerPolicy policy : rangerPolicies) {
+                atlasRangerService.deleteRangerPolicy(policy);
+            }
+        }
+    }
+
+    public void deletePurposePolicy(PurposeContext context) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("deletePurposePolicy");
+        context.setDeletePurposePolicy(true);
+
+        AtlasEntity purposePolicy = context.getPurposePolicy();
+
+        if(!POLICY_ENTITY_TYPE.equals(purposePolicy.getTypeName())) {
+            throw new AtlasBaseException(BAD_REQUEST, "Please provide entity of type " + POLICY_ENTITY_TYPE);
+        }
+
+        if(!purposePolicy.getStatus().equals(AtlasEntity.Status.ACTIVE)) {
+            LOG.info("Purpose policy with guid {} is already deleted/purged", purposePolicy.getGuid());
+            return;
+        }
+
+        List<RangerPolicy> provisionalRangerPolicies = purposePolicyToRangerPolicy(context);
+        updatePurposePolicies(context, provisionalRangerPolicies);
+
+        processDeletePolicy(context, provisionalRangerPolicies);
+
+        List<String> actions = getActions(purposePolicy);
+        if (actions.contains(ACCESS_ENTITY_READ)) {
+            aliasStore.updateAlias(context);
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private void processDeletePolicy(PurposeContext context, List<RangerPolicy> provisionalRangerPolicies) throws AtlasBaseException {
+        List<RangerPolicy> rangerPolicies = null;
+
+        if (CollectionUtils.isEmpty(provisionalRangerPolicies)) {
+            rangerPolicies = fetchRangerPoliciesByLabel(atlasRangerService,
+                    "tag",
+                    null,
+                    getPurposeLabel(context.getPurpose().getGuid()));
+        } else if (provisionalRangerPolicies.size() == 1) {
```

Comment:
Can there be a condition where size == 2 meaning that both policy items and data mask policy items are not empty and therefore both mask policy and ranger access policies are inside?

Reply:
to review

Reply:
A Purpose can have max 2 Ranger policies at a time

here actually there will never be a case when provisionalRangerPolicies size is <= 2 
Since this delete policy flow, this method will be called for only single policy delete which can be transformed into max 1 provisionalRangerPolicy